python/cudf/cudf/core/index.py

```diff
@@ -1925,6 +1991,31 @@ def get_dt_field(self, field):
         )
         return as_index(out_column, name=self.name)
 
+    def _clean_nulls_from_index(self):
+        """
+        Convert all null/NaT values(if any) in DatetimeIndex object
+        to `null` as a preprocessing step to `__repr__` methods.
+
+        This will involve changing type of Index object
+        to StringIndex but it is the responsibility of the `__repr__`
+        methods using this method to replace or handle representation
+        of the actual types correctly.
+
+        Note: We must not convert Datetime object to_pandas first as
+        this involves in expanded string representation of the
+        datetime values, instead we must perform `fillna` before
+        the type-cast operation.
+        """
+        if self._values.has_nulls:
+            return cudf.Index(
+                column.as_column(
+                    self.to_pandas().fillna("<NA>").astype("str")
+                ),
```

Comment:
This does a device --> host --> device. Can we either clean it on the device or leave it on the host?

Reply:
I'm thinking we should probably change all `_clean_nulls_from_...` methods to return only pandas objects instead of copying back to the device. Because essentially what we are doing with output of this API is do `to_pandas` on it. Thoughts?

Reply:
I think we should have a separation of concerns, where if this receives a GPU object, it may as well clean the nulls on the GPU and return a GPU object.

Reply:
This is now solved, the object will stay on the device. We shouldn't actually be converting to pandas before type-casting incase of datetime, as pandas supports only `datetime64[ns]` hence all the string outputs are getting truncated. Since we support multiple types `ns`, `ms`, `s`, etc., we should not be comparing with pandas here so added few tests for the same.